opt: introduce optimizer expression trees and build them from TypedExprs

RaduBerinde opened the pull request:
This change brings in a subset of
https://github.com/petermattis/opttoy/tree/master/v3

This change introduces:

 - the expr tree: cascades-style optimizers operate on expression
   trees which can represent both scalar and relational expressions;
   this is a departure from the way we represent expressions and
   statement (sem/tree) so we need a new tree structure.

 - scalar operators: initially, we focus only on scalar expressions.

 - building an expr tree from a sem/tree.TypedExpr.

 - opt version of logic tests

See the RFC in #19135 for more context on the optimizer.

This is the first step of an initial project related to the optimizer:
generating index constraints from scalar expressions. This will be a
rewrite of the current index constraint generation code (which has
many problems, see #6346).  Roughly, the existing
`makeIndexConstraints` will call into the optimizer with a `TypedExpr`
and the optimizer will return index constraints.

Release note: None

RaduBerinde committed:
opt: introduce optimizer expression trees and build them from TypedExprs

This change brings in a subset of
https://github.com/petermattis/opttoy/tree/master/v3

This change introduces:

 - the expr tree: cascades-style optimizers operate on expression
   trees which can represent both scalar and relational expressions;
   this is a departure from the way we represent expressions and
   statements (sem/tree) so we need a new tree structure.

 - scalar operators: initially, we focus only on scalar expressions.

 - building an expr tree from a sem/tree.TypedExpr.

 - opt version of logic tests

See the RFC in #19135 for more context on the optimizer.

This is the first step of an initial project related to the optimizer:
generating index constraints from scalar expressions. This will be a
rewrite of the current index constraint generation code (which has
many problems, see #6346).  Roughly, the existing
`makeIndexConstraints` will call into the optimizer with a `TypedExpr`
and the optimizer will return index constraints.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20557)
<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 9 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/expr.go, line 18 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-mkqGHdsOiGH_Cp8Hf:-L-mkqGHdsOiGH_Cp8Hg:bgg0bs5) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/expr.go#L18)):*
> ```Go
> 
> // expr is a unified interface for both relational and scalar expressions in a
> // query. TODO(radu): currently, it only supports scalar expressions.
> ```

Either pull in some of the additional commentary from `opttoty` or add another `TODO` about expanding the comments. The `opttoy` comments probably need to be rewritten. I'm fine with a `TODO`.

---

*[pkg/sql/opt/expr.go, line 19 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-mmJfmOVdeRPcj1RJJ:-L-mmJfmOVdeRPcj1RJK:b-urw5jn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/expr.go#L19)):*
> ```Go
> // expr is a unified interface for both relational and scalar expressions in a
> // query. TODO(radu): currently, it only supports scalar expressions.
> type expr struct {
> ```

I imagine we'll have to export `expr` at some point, though we can cross that bridge when we reach it.

---

*[pkg/sql/opt/opt_test.go, line 55 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-mjKMlZ8NNbOtXe47g:-L-mjKMlZ8NNbOtXe47h:b-gwfg7z) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/opt_test.go#L55)):*
> ```Go
> var (
> 	logicTestData    = flag.String("d", "testdata/[^.]*", "test data glob")
> 	rewriteTestFiles = flag.Bool("rewrite", false, "")
> ```

Time to copy over the help text from `rewrite-results-in-testfiles`. I was lazy in not doing that before.

---

*[pkg/sql/opt/tree_print.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-mjwHfMfXi0dDqn9Mp:-L-mjwHgzH4WWERir5iH:bgdpjd8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/tree_print.go#L30)):*
> ```Go
> //    |- child2
> //
> type treePrinter struct {
> ```

The name `treePrinter` is slightly confusing given the `sql/sem/tree` package. Perhaps this should be `exprPrinter`. Or perhaps `exprDebugPrinter`, though that is a bit verbose. 

---

*[pkg/sql/opt/tree_print.go, line 34 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-mjWCenPm7CzzIdEB7:-L-mjWCenPm7CzzIdEB8:bdfu3vw) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/tree_print.go#L34)):*
> ```Go
> 
> 	// We maintain the rows accumulated so far.
> 	// When a new child is adedd (e.g. child2 above), we may have to go back up
> ```

`s/adedd/added/g`

---

*[pkg/sql/opt/testdata/build-scalar, line 11 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-mlS4xwiQr2YliHrv4:-L-mlS4xwiQr2YliHrv5:bsjjuoi) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/testdata/build-scalar#L11)):*
> ```
> const (1) (type: int)
> ```

Are the 2 blank lines intentional? If yes, you're not consistent below. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-:-L-mmF7v0XgXkeYGuuao:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
I'm sorry but the commit message doesn't cut it. 
6 months down the line, or the next batch of hires, will not understand anything about the backstory.
The commit message must explain what is going on *without appealing to implicit context*

So:

- expand the commit title
- summarize what the different types are for in the commit message
- remind the reader where this is going

RaduBerinde commented:
> summarize what the different types are for in the commit message

What types? The types in the code should be documented in the code, not in commit messages.

> remind the reader where this is going

The commit message already includes this paragraph:
```
The plan is to use this framework to rewrite index constraint generation
(#6346). The sql code will pass a `TypedExpr` to the optimizer and have it
compute the index constraints.
```

petermattis commented:
> The commit message must explain what is going on without appealing to implicit context

That's a high bar to clear and not one that I would personally impose.

knz commented:
Okay let me try that for you:

```
sql: introduce new data structures to represent logical plans

This patch aims introduces new data structures that will facilitate the introduction of a new logical plan optimization infrastructure.

This is achieved by introducing:
- a new generic `expr` type intended to represent any node in a logical plan
- a concrete enumeration of constant values that denote `expr` sub-types, that differentiate between logical plan operators, sufficient to encode *scalar* expressions
- to exercise the code, a recursive function `buildScalar` which takes a `tree.TypedExpr` tree and converts it to an `expr` tree
- a couple of a tests that check these new features.

See the RFC [...ref...] for context.

Release note: None
```

knz commented:
Also it is a mistake to place the `expr` struct definitions in the `opt` package.

It should go to `sql/sem/expr`.

knz commented:
Note that I insist that *a competent reviewer must be able to check and understand this code without looking at opttoy and even know that it exists*. There are a few more shortcomings, I have more comments coming soon.

petermattis commented:
> Note that I insist that a competent reviewer must be able to check and understand this code without looking at opttoy and even know that it exists. There are a few more shortcomings, I have more comments coming soon.

Where are we going to find such an engineer? Or are you going to wipe knowledge of `opttoy` from your memory?

There are definite shortcomings to the code and some design decisions that won't be clear without additional context. IMO, the hurdles you appear to be putting up will not be justified by the benefit (if any) they provide. On the downsides of the hurdles are significant. Clearly you feel very strongly about this and I imagine discussing in PR comments will be fraught. Let's find time to discuss in person.

knz commented:
Reviewed 8 of 9 files at r1.
Review status: 8 of 9 files reviewed at latest revision, 21 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/build.go, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-mt17QCRooX74yQwW4:-L-mt17RilxbZzmWX64e:b-ajrar) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/build.go#L1)):*
> ```Go
> // Copyright 2017 The Cockroach Authors.
> ```

This file belongs to a new package `sql/sem/expr`.

---

*[pkg/sql/opt/build.go, line 80 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-mt5eVIJUtSVgLjZZy:-L-mt5eVIJUtSVgLjZZz:bnyddtc) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/build.go#L80)):*
> ```Go
> }
> 
> const exprAllocChunk = 16
> ```

Document these two constants.

---

*[pkg/sql/opt/expr.go, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-mtE4_YHA-3uowEhIc:-L-mtE4_YHA-3uowEhId:b-ajrar) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/expr.go#L1)):*
> ```Go
> // Copyright 2017 The Cockroach Authors.
> ```

This file belongs to a new package `sql/sem/expr`.

---

*[pkg/sql/opt/expr.go, line 24 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-mtHDO3hCHyHZ-myyO:-L-mtHDO3hCHyHZ-myyP:b8gcih4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/expr.go#L24)):*
> ```Go
> 	// dependent.
> 	children []*expr
> 	// Scalar properties.
> ```

Extend the comment to remind the reader what are scalar properties.

---

*[pkg/sql/opt/misc.go, line 19 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-mtNZ1M7kRMF6JJktT:-L-mtNZ1M7kRMF6JJktU:bs7117d) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/misc.go#L19)):*
> ```Go
> import "fmt"
> 
> func panicf(format string, args ...interface{}) {
> ```

This function should be added in a separate commit.

---

*[pkg/sql/opt/operator.go, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-mtQrAu7PdQL7t6kug:-L-mtQrAu7PdQL7t6kuh:b-ajrar) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/operator.go#L1)):*
> ```Go
> // Copyright 2017 The Cockroach Authors.
> ```

This file belongs to a new package `sql/sem/expr`.

---

*[pkg/sql/opt/operator.go, line 26 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-mtWn35Ml6NaBRxxqO:-L-mtWn35Ml6NaBRxxqP:b-2dchxx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/operator.go#L26)):*
> ```Go
> 	// TODO(radu): no relational operators yet.
> 
> 	// Scalar operators
> ```

Every operator below `exists` is rather self-explanatory. The first 4 are non-trivial however and should be further explained in comments.

---

*[pkg/sql/opt/operator.go, line 82 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-mtckaYoLoMBWuQfII:-L-mtckaYoLoMBWuQfIJ:b-hp3g2d) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/operator.go#L82)):*
> ```Go
> 	unaryComplementOp
> 
> 	functionOp
> ```

This is a misnomer.

Make this `callOp` or `functionCallOp`

---

*[pkg/sql/opt/operator.go, line 87 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-mtjWn0gtwV6wjkSRK:-L-mtjWn0gtwV6wjkSRL:b-vl8rne) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/operator.go#L87)):*
> ```Go
> )
> 
> type operatorInfo struct {
> ```

Detail what this struct is for in comments, as well as its members.

---

*[pkg/sql/opt/operator.go, line 92 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-mtnJ_FFjb-xiQBVK7:-L-mtnJaHy87BsxFmQyb:bdww09f) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/operator.go#L92)):*
> ```Go
> }
> 
> var operatorTab = [numOperators]operatorInfo{
> ```

Explain this in comments.

---

*[pkg/sql/opt/operator.go, line 103 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-mtoMoxzeqK9f_z4ZC:-L-mtoMoxzeqK9f_z4ZD:b918u1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/operator.go#L103)):*
> ```Go
> }
> 
> func registerOperator(op operator, name string, class operatorClass) {
> ```

Explain what this does and why it exists.

---

*[pkg/sql/opt/scalar.go, line 26 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-mu-0EdFT-c2frzuTU:-L-mu-0EdFT-c2frzuTV:b-1fwf3f) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/scalar.go#L26)):*
> ```Go
> 
> func init() {
> 	registerOperator(variableOp, "variable", scalarClass{})
> ```

You could define a scalar-specific `operatorTab` here and have it merged into the global one. I don't think that calling `registerOperator` many times in the proper way to structure this code.

---

*[pkg/sql/opt/scalar.go, line 40 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-mu8Rp0D9XsKMbYz8C:-L-mu8Rp0D9XsKMbYz8D:b-uhwz48) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/scalar.go#L40)):*
> ```Go
> 	registerOperator(geOp, "ge", scalarClass{})
> 	registerOperator(neOp, "ne", scalarClass{})
> 	registerOperator(inOp, "IN", scalarClass{})
> ```

the capitals here and below are inconsistent with the other operators.

This needs to either be made consistent, or extensively documented both here and the commit message to justify the inconsistency.

---

*[pkg/sql/opt/scalar.go, line 99 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-muJy9ALVlle0wn_z_:-L-muJyA-ufoEurto_TV:bkas7a5) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/scalar.go#L99)):*
> ```Go
> }
> 
> func initConstExpr(e *expr, datum tree.Datum) {
> ```

These various constructors must be commented / explained further.

---

*[pkg/sql/opt/tree_print.go, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-muN_xRjJZ2uRTgtn6:-L-muN_xRjJZ2uRTgtn7:b-ajrar) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/tree_print.go#L1)):*
> ```Go
> // Copyright 2017 The Cockroach Authors.
> ```

This file belongs to a new package `sql/sem/expr`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557)*
<!-- Sent from Reviewable.io -->

knz commented:
@petermattis it's very clear to me that any of Jordan, Nathan, Toby or Justin must be able to review and understand this code without looking at opttoy. If they don't (or can't) we have a serious organisatory problem.

RaduBerinde commented:
I am happy to add more comments and make things more understandable (within reason).

> Also it is a mistake to place the expr struct definitions in the opt package. It should go to sql/sem/expr.

I don't yet have a picture of how this should be divided into packages; I would rather keep things in one place for now and move things out later (not to mention that having `expr.Expr` and `tree.Expr` is confusing, so I'd rather not start with that). Note that as far as the initial constraint-generation project is concerned, expr can be private to opt.

RaduBerinde commented:
Review status: 8 of 9 files reviewed at latest revision, 21 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/misc.go, line 19 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-mtNZ1M7kRMF6JJktT:-L-mvpZHRufFTluS6sx4:bqgnr87) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/misc.go#L19)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This function should be added in a separate commit.
</blockquote></details>

Huh? These are all new files, and you chose this one-line-wrapper as something that needs its own commit?

---

*[pkg/sql/opt/tree_print.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-mjwHfMfXi0dDqn9Mp:-L-mw0hRaBpCGJG1GO2l:bgkdt89) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ed8c89251395492976e60e5149a6fa748bd1ad5/pkg/sql/opt/tree_print.go#L30)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The name `treePrinter` is slightly confusing given the `sql/sem/tree` package. Perhaps this should be `exprPrinter`. Or perhaps `exprDebugPrinter`, though that is a bit verbose. 
</blockquote></details>

I was thinking of moving this to `util` as it's not specific to expressions and could be useful for other things (like EXPLAIN).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557)*
<!-- Sent from Reviewable.io -->

knz commented:
ok we can think about separate packages later, I agree.

RaduBerinde commented:
Updated the commit message, PTAL. I will address the rest of the comments later today.

---

Review status: 8 of 9 files reviewed at latest revision, 22 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-:-L-mwvN5BXuLPwS-mzKu:b3fegij)*
<!-- Sent from Reviewable.io -->

knz commented:
FWIW I have read the code and I recognize in it what the RFC defines and that it aligns with the overall directions.  @jordanlewis can you do quality control on it?

RaduBerinde committed:
util: add tree printer package

Adding a helper that can be used to print hierarchical structures (to be used
with optimizer expressions).

Release note: None

petermattis commented:
Review status: 0 of 9 files reviewed at latest revision, 22 unresolved discussions.

---

*[pkg/sql/opt/build.go, line 145 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-qhvadkxFg5VE3rO69:-L-qhvadkxFg5VE3rO6A:b-s5elvv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d49013e84fd46775e9e1a0095130cc25425bde3a/pkg/sql/opt/build.go#L145)):*
> ```Go
> 
> 	case *tree.FuncExpr:
> 		def, err := t.Func.Resolve(tree.SearchPath{})
> ```

Isn't the function already resolved by this point? I would have imagined we would need to resolve the function before typing. Answering my own question: looks like this is just the method you call to get the resolved function. I suppose we could add a `ResolvedFunc` method (similar to `ResolvedType`) for clarity, but doesn't seem too important.

---

*[pkg/sql/opt/scalar.go, line 86 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-qh1ndOalz0gc2S4AP:-L-qh1ndOalz0gc2S4AQ:b-ln5cms) ([raw file](https://github.com/cockroachdb/cockroach/blob/d49013e84fd46775e9e1a0095130cc25425bde3a/pkg/sql/opt/scalar.go#L86)):*
> ```Go
> // scalarProps are properties specific to scalar operators.
> type scalarProps struct {
> 	typ types.T
> ```

Let's comment this field as long as you're on a commenting spree.

---

*[pkg/sql/opt/scalar.go, line 105 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-qh7tEI8Rg5Uy0-XS-:-L-qh7tEI8Rg5Uy0-XS0:bhw7s68) ([raw file](https://github.com/cockroachdb/cockroach/blob/d49013e84fd46775e9e1a0095130cc25425bde3a/pkg/sql/opt/scalar.go#L105)):*
> ```Go
> 
> // The following initializers are called on an already allocated expression
> // node. The scalarProps are be initialized separately.
> ```

`s/are be/will be/g`

---

*[pkg/sql/opt/testdata/build-scalar, line 45 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-qfzUpHLZCPLLMyGeJ:-L-qfzUqEoouvpWkscLg:b-rbkx1v) ([raw file](https://github.com/cockroachdb/cockroach/blob/d49013e84fd46775e9e1a0095130cc25425bde3a/pkg/sql/opt/testdata/build-scalar#L45)):*
> ```
>       │    │    │    │    ├── const (5) (type: int)
>       │    │    ├──  │ (t │e: bool)
>       │    │    │    └──  │puts
> ```

Something seems messed up here. The about should read `inputs`. And the previous line also seems to have a problem where we have `t │e` instead of `type`.

---

*[pkg/sql/opt/testdata/build-scalar, line 51 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-qftBuip6_wWdUqSz-:-L-qftBvAnnEni3QnKf2:b-ywatp6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d49013e84fd46775e9e1a0095130cc25425bde3a/pkg/sql/opt/testdata/build-scalar#L51)):*
> ```
>            └──  │puts
>                 ├── variable (1) (type: int)
>                 └── const (4) (type: int)
> ```

Very pretty!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
> @jordanlewis can you do quality control on it?

@jordanlewis To expand on this request - it would be useful to get your review as someone who hasn't looked at opttoy in detail. There may be things which seem "obvious" to us but are not. There is no rush though. Thanks!

RaduBerinde commented:
Review status: 0 of 9 files reviewed at latest revision, 22 unresolved discussions, some commit checks pending.

---

*[pkg/sql/opt/build.go, line 145 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-qhvadkxFg5VE3rO69:-L-qngIhEp0GhOSxwKsT:b8pcc7t) ([raw file](https://github.com/cockroachdb/cockroach/blob/d49013e84fd46775e9e1a0095130cc25425bde3a/pkg/sql/opt/build.go#L145)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Isn't the function already resolved by this point? I would have imagined we would need to resolve the function before typing. Answering my own question: looks like this is just the method you call to get the resolved function. I suppose we could add a `ResolvedFunc` method (similar to `ResolvedType`) for clarity, but doesn't seem too important.
</blockquote></details>

We don't hold on to the FunctionDefinition in `FuncExpr` so I don't know how to return one without resolving it again.

---

*[pkg/sql/opt/scalar.go, line 86 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-qh1ndOalz0gc2S4AP:-L-qkcuAYMHg8BrcOUT8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d49013e84fd46775e9e1a0095130cc25425bde3a/pkg/sql/opt/scalar.go#L86)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Let's comment this field as long as you're on a commenting spree.
</blockquote></details>

Done.

---

*[pkg/sql/opt/scalar.go, line 105 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-qh7tEI8Rg5Uy0-XS-:-L-qk0V9ZjwyJ__fFRKH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d49013e84fd46775e9e1a0095130cc25425bde3a/pkg/sql/opt/scalar.go#L105)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/are be/will be/g`
</blockquote></details>

Done.

---

*[pkg/sql/opt/testdata/build-scalar, line 45 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-qfzUpHLZCPLLMyGeJ:-L-qkssTd4IOlbkb2kSj:b4rftji) ([raw file](https://github.com/cockroachdb/cockroach/blob/d49013e84fd46775e9e1a0095130cc25425bde3a/pkg/sql/opt/testdata/build-scalar#L45)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Something seems messed up here. The about should read `inputs`. And the previous line also seems to have a problem where we have `t │e` instead of `type`.
</blockquote></details>

Oh, this is bad, thank you for noticing! Fixed (and improved the treeprinter test).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 9 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/build.go, line 145 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-qhvadkxFg5VE3rO69:-L-rb7fNEAHDL0-cpWY4:b-jknrto) ([raw file](https://github.com/cockroachdb/cockroach/blob/d49013e84fd46775e9e1a0095130cc25425bde3a/pkg/sql/opt/build.go#L145)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We don't hold on to the FunctionDefinition in `FuncExpr` so I don't know how to return one without resolving it again.
</blockquote></details>

We do hold onto the `FunctionDefinition`. `tree.ResolvableFunctionReference.Resolve()` caches the function resolution. My point was that is is somewhat confusing that `FuncExpr.Resolve` is using this cached data and perhaps we should add a `FuncExpr.ResolvedFunc` that requires the function to already be resolved. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 10 files reviewed at latest revision, 17 unresolved discussions.

---

*[pkg/sql/opt/build.go, line 145 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-qhvadkxFg5VE3rO69:-L-ru59KSYqN17HXPcg4:b-ymbwsq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d49013e84fd46775e9e1a0095130cc25425bde3a/pkg/sql/opt/build.go#L145)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

We do hold onto the `FunctionDefinition`. `tree.ResolvableFunctionReference.Resolve()` caches the function resolution. My point was that is is somewhat confusing that `FuncExpr.Resolve` is using this cached data and perhaps we should add a `FuncExpr.ResolvedFunc` that requires the function to already be resolved. 
</blockquote></details>

Ah, got it. Done, PTAL

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
@RaduBerinde @knz I am planning to review this soon!

rytaft commented:
Reviewed 2 of 9 files at r4, 7 of 8 files at r5.
Review status: 9 of 10 files reviewed at latest revision, 26 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/build.go, line 76 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05mM5Nv1CRAvmQJVJS:-L05mM5Nv1CRAvmQJVJT:blc6gv4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/sql/opt/build.go#L76)):*
> ```Go
> // We allocate *scalarProps and *expr in chunks of these sizes.
> const exprAllocChunk = 16
> const scalarPropsAllocChunk = 16
> ```

Where did 16 come from? Why is that a good choice?

---

*[pkg/sql/opt/expr.go, line 61 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05osNYY664JWdudy6I:-L05osNYY664JWdudy6J:b-bb0zxl) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/sql/opt/expr.go#L61)):*
> ```Go
> }
> 
> // format is part of the operatorInfo interface.
> ```

Don't you mean operatorClass interface?

---

*[pkg/sql/opt/misc.go, line 19 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05nZILS0OqOLhFsCg7:-L05nZILS0OqOLhFsCg8:b-f1u0iy) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/sql/opt/misc.go#L19)):*
> ```Go
> import "fmt"
> 
> func panicf(format string, args ...interface{}) {
> ```

This seems like a basic function that would be useful for other packages too.  Maybe it belongs in util?  (Maybe this was Raphael's intention in asking for a separate commit?)

Anyway, misc.go isn't a very informative name....

---

*[pkg/sql/opt/opt_test.go, line 18 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05tZU1GZaf9a2xC7eL:-L05tZU1GZaf9a2xC7eM:b2mq9rq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/sql/opt/opt_test.go#L18)):*
> ```Go
> 
> // This file is home to TestOpt, which is similar to the logic tests, except it
> // is used for optimizer-specific testcases.
> ```

Why not just augment logic_test.go with the build-scalar command?  It seems like a lot of this code is repeated from that file.

---

*[pkg/sql/opt/opt_test.go, line 85 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05qRBo9WdH9w7GfonE:-L05qRBo9WdH9w7GfonF:b-tggezd) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/sql/opt/opt_test.go#L85)):*
> ```Go
> type testdata struct {
> 	pos      string // file and line number
> 	cmd      string // exec, query, ...
> ```

I thought currently only build-scalar was supported (this comment should be fixed)

---

*[pkg/sql/opt/scalar.go, line 115 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05udTRfcrerPqu-e_4:-L05udTRfcrerPqu-e_5:b-reck2f) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/sql/opt/scalar.go#L115)):*
> ```Go
> }
> 
> // initFunctionCallExpra initializes a functionCallOp expression node.
> ```

`initFunctionCallExpra` -> `initFunctionCallExpr`

---

*[pkg/sql/opt/scalar.go, line 135 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05v2pc8uVg2cNjLdAE:-L05v2pc8uVg2cNjLdAF:bhddxj7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/sql/opt/scalar.go#L135)):*
> ```Go
> 
> // initVariableExpr initializes a variableOp expression node.
> // TODO(radu): currently variableOp refers to a column index (derived from an
> ```

What should it refer to instead?

---

*[pkg/sql/opt/testdata/build-scalar, line 51 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L-qftBuip6_wWdUqSz-:-L05vQ5CtU_kQhwK2qzu:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/d49013e84fd46775e9e1a0095130cc25425bde3a/pkg/sql/opt/testdata/build-scalar#L51)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Very pretty!
</blockquote></details>

+1

---

*[pkg/sql/opt/testdata/build-scalar, line 6 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05vLGXvUEsGymO4fwh:-L05vLGXvUEsGymO4fwi:bnvhp16) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/sql/opt/testdata/build-scalar#L6)):*
> ```
> const (1) (type: int)
> 
> build-scalar
> ```

Why is this test case repeated?

---

*[pkg/util/treeprinter/tree_printer.go, line 52 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05w33q-ZDVMOUIMalq:-L05w33q-ZDVMOUIMalr:b1x6510) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/util/treeprinter/tree_printer.go#L52)):*
> ```Go
> //    ├── child-2
> //    │    └── grandchild
> //    └── child-1
> ```

`child-1` -> `child-3`

---

*[pkg/util/treeprinter/tree_printer.go, line 114 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05wqw6DPp4zDqSMqkt:-L05wqw6DPp4zDqSMqku:b-sqsebt) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/util/treeprinter/tree_printer.go#L114)):*
> ```Go
> 	n.tree.rows = append(n.tree.rows, row)
> 
> 	// Return a TreePrinter that can be used for children of this nodes.
> ```

`nodes` -> `node`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTR!

---

Review status: 1 of 9 files reviewed at latest revision, 26 unresolved discussions.

---

*[pkg/sql/opt/build.go, line 76 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05mM5Nv1CRAvmQJVJS:-L06-3PIyhPOw5mz26J8:bh8r1h1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/sql/opt/build.go#L76)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Where did 16 come from? Why is that a good choice?
</blockquote></details>

It's an arbitrary choice, just amortizing some allocations.

---

*[pkg/sql/opt/expr.go, line 61 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05osNYY664JWdudy6I:-L06-VK3IgKzIQmV0rvu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/sql/opt/expr.go#L61)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Don't you mean operatorClass interface?
</blockquote></details>

Done.

---

*[pkg/sql/opt/opt_test.go, line 18 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05tZU1GZaf9a2xC7eL:-L060BGEQRulyaFqH_Ws:b-q0f0n9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/sql/opt/opt_test.go#L18)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Why not just augment logic_test.go with the build-scalar command?  It seems like a lot of this code is repeated from that file.
</blockquote></details>

First, logic test is in a separate package and this test should be able to use unexported `opt` functionality. Second, logic-test always runs SQL, not clear adding this kind of functionality there is a good idea.

---

*[pkg/sql/opt/opt_test.go, line 85 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05qRBo9WdH9w7GfonE:-L060vg7DEQp4whniX98:bdsekjz) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/sql/opt/opt_test.go#L85)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I thought currently only build-scalar was supported (this comment should be fixed)
</blockquote></details>

Removed the comment (the supported commands are described in the beginning).

---

*[pkg/sql/opt/scalar.go, line 115 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05udTRfcrerPqu-e_4:-L061274FOCq9lvvHtZU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/sql/opt/scalar.go#L115)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

`initFunctionCallExpra` -> `initFunctionCallExpr`
</blockquote></details>

Done.

---

*[pkg/sql/opt/scalar.go, line 135 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05v2pc8uVg2cNjLdAE:-L0614dKiLncYfrtresj:bklmjju) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/sql/opt/scalar.go#L135)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

What should it refer to instead?
</blockquote></details>

Removed the TODO

---

*[pkg/sql/opt/testdata/build-scalar, line 6 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05vLGXvUEsGymO4fwh:-L061UZwVtpbEVQWc4nI:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/sql/opt/testdata/build-scalar#L6)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Why is this test case repeated?
</blockquote></details>

Fixed

---

*[pkg/util/treeprinter/tree_printer.go, line 52 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05w33q-ZDVMOUIMalq:-L0621l41ajpy0c-gVGp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/util/treeprinter/tree_printer.go#L52)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

`child-1` -> `child-3`
</blockquote></details>

Done.

---

*[pkg/util/treeprinter/tree_printer.go, line 114 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05wqw6DPp4zDqSMqkt:-L0623iq13RobtdT5E_f:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/util/treeprinter/tree_printer.go#L114)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

`nodes` -> `node`
</blockquote></details>

Done.

---

*[pkg/sql/opt/misc.go, line 19 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05nZILS0OqOLhFsCg7:-L06-n46yQoE9rEjrS-O:bsyk59b) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/sql/opt/misc.go#L19)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

This seems like a basic function that would be useful for other packages too.  Maybe it belongs in util?  (Maybe this was Raphael's intention in asking for a separate commit?)

Anyway, misc.go isn't a very informative name....
</blockquote></details>

Sigh.. removed it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-:-L0625MWDu7D6k-HaWgO:b-814g8k)*
<!-- Sent from Reviewable.io -->

rytaft commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 9 files at r6, 7 of 7 files at r7.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks pending.

---

*[pkg/sql/opt/scalar.go, line 135 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05v2pc8uVg2cNjLdAE:-L063kYJh6Isqrj4qQvS:b-auz8t9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/sql/opt/scalar.go#L135)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Removed the TODO
</blockquote></details>

It would help to keep the info about what `index` means, though.

---

*[pkg/util/treeprinter/tree_printer.go, line 52 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05w33q-ZDVMOUIMalq:-L0649Yjzph2IY-Zni3u:b-a0g18j) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/util/treeprinter/tree_printer.go#L52)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Done.
</blockquote></details>

you changed it to `child-4` instead of `child-3` (at least make it consistent with the code above)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-:-L065oazjuZU6wHGfhR3:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 1 of 9 files reviewed at latest revision, 18 unresolved discussions.

---

*[pkg/sql/opt/scalar.go, line 135 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05v2pc8uVg2cNjLdAE:-L067IwyRTchbzVIwTSe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/sql/opt/scalar.go#L135)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

It would help to keep the info about what `index` means, though.
</blockquote></details>

Done.

---

*[pkg/util/treeprinter/tree_printer.go, line 52 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L05w33q-ZDVMOUIMalq:-L066wfc5L3eRWpE-vyF:b-rst3ay) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d7f79bea70e38a74e5f37ddbb0806f94f43bb3d/pkg/util/treeprinter/tree_printer.go#L52)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

you changed it to `child-4` instead of `child-3` (at least make it consistent with the code above)
</blockquote></details>

Sorry, fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 1 of 8 files at r8, 7 of 7 files at r9.
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 9 files at r4, 8 of 8 files at r8, 7 of 7 files at r9.
Review status: all files reviewed at latest revision, 21 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/expr.go, line 20 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L06VQvjQ9xL-bQci9Aq:-L06VQvjQ9xL-bQci9Ar:b-gk6h5o) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef09422a1c4d69ce57d2c57444d39251457468ab/pkg/sql/opt/expr.go#L20)):*
> ```Go
> 
> // expr implements the node of a unified expressions tree for both relational
> // and scalar expressions in a query.
> ```

I would benefit from examples here as an outsider. I imagine a scalar expression is what our current "expression tree" represents. What is a relational expression precisely? Does it correspond to what we call a "statement" or is it more limited than that? Can you give some examples? 

I see below that you're planning to do that later - that's okay then.

---

*[pkg/sql/opt/expr.go, line 22 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L06WDyaimWoAVn98Hg_:-L06WDyaimWoAVn98Hga:bhee1d5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef09422a1c4d69ce57d2c57444d39251457468ab/pkg/sql/opt/expr.go#L22)):*
> ```Go
> // and scalar expressions in a query.
> //
> // Expressions have optional inputs. Expressions also maintain properties. For
> ```

Also would be nice to have examples here -of course they maintain properties. Nearly everything in software does! What is an expression property?

---

*[pkg/sql/opt/expr.go, line 37 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L06WSDY-Ont5K0Dbi0z:-L06WSDY-Ont5K0Dbi1-:bxy6qss) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef09422a1c4d69ce57d2c57444d39251457468ab/pkg/sql/opt/expr.go#L37)):*
> ```Go
> 	// Scalar properties (properties that pertain only to scalar operators).
> 	scalarProps *scalarProps
> 	// Operator-dependent data used by this expression. For example, constOp
> ```

This is good - could you add a For example line to the other fields?

When code gets this generic, I think it's very helpful to provide copious examples.

---

*[pkg/sql/opt/operator.go, line 75 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L06XMyhDwSsOy6Eq9OI:-L06XMyhDwSsOy6Eq9OJ:b-bstnds) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef09422a1c4d69ce57d2c57444d39251457468ab/pkg/sql/opt/operator.go#L75)):*
> ```Go
> 	allOp
> 
> 	bitandOp
> ```

nit: For consistency with the camel cased ones above consider `bitAndOp` etc

---

*[pkg/sql/opt/operator.go, line 95 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L06XIOhy0fcxIBZJf9F:-L06XIOhy0fcxIBZJf9G:b-phy9qf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef09422a1c4d69ce57d2c57444d39251457468ab/pkg/sql/opt/operator.go#L95)):*
> ```Go
> 	functionCallOp
> 
> 	numOperators
> ```

What is `numOperators`?

Oh now I see it is around to get the number of operators! Please add a comment that this must come last.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-:-L06YC3mqBzFevVsxfJC:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTRs!

---

Review status: all files reviewed at latest revision, 21 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/expr.go, line 20 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L06VQvjQ9xL-bQci9Aq:-L09if3LPX4OHUfnlSUL:bmxl1qz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef09422a1c4d69ce57d2c57444d39251457468ab/pkg/sql/opt/expr.go#L20)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I would benefit from examples here as an outsider. I imagine a scalar expression is what our current "expression tree" represents. What is a relational expression precisely? Does it correspond to what we call a "statement" or is it more limited than that? Can you give some examples? 

I see below that you're planning to do that later - that's okay then.

</blockquote></details>

Yeah, I will add more information when we support relational expressions. I figured it'd be confusing to explain things that aren't implemented.

---

*[pkg/sql/opt/expr.go, line 22 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L06WDyaimWoAVn98Hg_:-L09k-ay8n1eVPV3kfk_:b-r20e4x) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef09422a1c4d69ce57d2c57444d39251457468ab/pkg/sql/opt/expr.go#L22)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Also would be nice to have examples here -of course they maintain properties. Nearly everything in software does! What is an expression property?
</blockquote></details>

Fixed the comment a bit, PTAL. I'm not sure what more to say. There will be more information when we support more kinds of expressions.

---

*[pkg/sql/opt/expr.go, line 37 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L06WSDY-Ont5K0Dbi0z:-L09kiG5KMhLRIgk_dqM:b-718w14) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef09422a1c4d69ce57d2c57444d39251457468ab/pkg/sql/opt/expr.go#L37)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This is good - could you add a For example line to the other fields?

When code gets this generic, I think it's very helpful to provide copious examples.
</blockquote></details>

Added a line for children. The scalarProps type is pretty well defined.

---

*[pkg/sql/opt/operator.go, line 75 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L06XMyhDwSsOy6Eq9OI:-L09kzoLbZ1Yu7Wm7njh:b92711r) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef09422a1c4d69ce57d2c57444d39251457468ab/pkg/sql/opt/operator.go#L75)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

nit: For consistency with the camel cased ones above consider `bitAndOp` etc
</blockquote></details>

This is consistent with `tree.Bitand` etc. I guess we treat them as one word.

---

*[pkg/sql/opt/operator.go, line 95 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-L06XIOhy0fcxIBZJf9F:-L09l7AR0Hk71G9jtagh:bhsszbp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef09422a1c4d69ce57d2c57444d39251457468ab/pkg/sql/opt/operator.go#L95)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

What is `numOperators`?

Oh now I see it is around to get the number of operators! Please add a comment that this must come last.
</blockquote></details>

Done. This is a very common pattern.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20557#-:-L09lthQXC2n71Ewu4g0:bz5mair)*
<!-- Sent from Reviewable.io -->
